### ISSUE TITLE
Migrate to TypeScript

### ISSUE THREAD

**Issue by queicherius** [queicherius/gw2api-client#64]

**Comment by darthmaim** [comment]
This would be useful so other devs don't need to define files like https://github.com/Archomeda/gw2-console/blob/master/types/gw2api-client.d.ts

**Comment by ogrady** [comment]
I am currently working on a migration, but obviously, the process is quite cumbersome as defining all possible response types takes a lot of time.

What's the scope of this issue? Just a basic migration with several result types available and some possible still being `any`, or a full migration with all types being present?

**Comment by queicherius** [comment]
> I am currently working on a migration, but obviously, the process is quite cumbersome as defining all possible response types takes a lot of time.

Yeah, I tried auto-generating some types before as well, but the quality of those was pretty meh.

> What's the scope of this issue? Just a basic migration with several result types available and some possible still being `any`, or a full migration with all types being present?

I think the goal is definitely full type support with all types. But I'd be more than happy to merge a PR with some types, it doesnt have to be all of them. 👍 Maybe even start with something small, so we can align the style.

**Comment by ogrady** [comment]
Okay, I'll just poke a bit around and see how far I can get. It may actually require slight adjustments in the API if you'd like to have the types as strict as possible, as the rolling schema version is hard to express via types.
I will open a PR as soon as I have a somewhat stable state. Just wanted to make sure the issue is still relevant beforehand. :)

**Comment by craimasjien** [comment]
Any update on this @ogrady @queicherius ?

**Comment by ogrady** [comment]
Hi everyone,

I defined [most endpoints](https://github.com/ogrady/gw2api-client/issues/1) manually in my fork, but then got swamped with other projects and irl stuff. I hope to continue at some point, maybe around the holidays, but there is still [a truckload of work](https://github.com/ogrady/gw2api-client/issues) to do. 
I started semi-automating the process of reading the types from the wiki, which brought more consistency into the types. But the ones I did before that would require at least a second round to double-check them and carry over the comments from the wiki.
After that, we still need to assign the types to each endpoint, add testing, etc.

tl;dr: I have made good progress but had to pause.

**Comment by sickhippie** [comment]
To anyone watching out for this one, I talked with @ogrady for a bit today and I'm going to take this issue over from the state of the latest commit on the open PR to try to wrap up the rest of it over the next few weeks (time permitting of course).
